### PR TITLE
Ensure that Scene::toImage renders texture backed images.

### PR DESCRIPTION
**Pull request by chinmaygarde**
TL;DR: Offscreen surface is created on the render thread and device to host
transfer performed there before task completion on the UI thread.

While attempting to snapshot layer trees, the engine was attempting to use the
IO thread context. The reasoning was that this would be safe to do because any
textures uploaded to the GPU as a result of async texture upload would have
originated from this context and hence the handles would be valid in either
context. As it turns out, while the handles are valid, Skia does not support
this use-case because cross-context images transfer ownership of the image from
one context to another. So, when we made the hop from the UI thread to the IO
thread (for snapshotting), if either the UI or GPU threads released the last
reference to the texture backed image, the image would be invalid. This led to
such images being absent from the layer tree snapshot.

Simply referencing the images as they are being used on the IO thread is not
sufficient because accessing images on one context after their ownership has
already been transferred to another is not safe behavior (from Skia's
perspective, the handles are still valid in the sharegroup).

To work around these issues, it was decided that an offscreen render target
would be created on the render thread. The color attachment of this render
target could then be transferred as a cross context image to the IO thread for
the device to host tranfer.

Again, this is currently not quite possible because the only way to create
cross context images is from encoded data. Till Skia exposes the functionality
to create cross-context images from textures in one context, we do a device to
host transfer on the GPU thread. The side effect of this is that this is now
part of the frame workload (image compression, which dominate the wall time,
is still done of the IO thread).

A minor side effect of this patch is that the GPU latch needs to be waited on
before the UI thread tasks can be completed before shell initialization.

Fixes https://github.com/flutter/flutter/issues/17687

### CODE REVIEW

**Comment by chinmaygarde** [comment]
Tested on GPU and software rendering environments. Traces showing the workload on the different threads [is attached](https://github.com/flutter/engine/files/2503816/RenderThreadToScene.json.zip) (debug).

Thanks to @andreidiaconu  for the reduced test case:
img_0004

File: shell/common/rasterizer.cc [hunk 1]
```diff
@@ -89,6 +93,53 @@ void Rasterizer::Draw(
...
+    TRACE_EVENT0("flutter", "MakeDeviceSnpshot");
+    device_snapshot = surface->makeImageSnapshot();
+  }
+
+  if (device_snapshot == nullptr) {
+    return nullptr;
+  }
+
+  {
+    TRACE_EVENT0("flutter", "DeviceHostTranfer");
```
Comment (jason-simmons) [inline shell/common/rasterizer.cc]: typo: "transfer"

File: shell/common/rasterizer.cc [hunk 2]
```diff
@@ -89,6 +93,53 @@ void Rasterizer::Draw(
...
+    return nullptr;
+  }
+
+  surface->getCanvas()->drawPicture(picture.get());
+
+  surface->getCanvas()->flush();
+
+  sk_sp<SkImage> device_snapshot;
+  {
+    TRACE_EVENT0("flutter", "MakeDeviceSnpshot");
```
Comment (jason-simmons) [inline shell/common/rasterizer.cc]: typo: "snapshot"

File: lib/ui/compositing/scene.cc [hunk 3]
```diff
@@ -110,67 +70,70 @@ Dart_Handle Scene::toImage(uint32_t width,
...
+      tonic::DartInvoke(image_callback->Get(), {raw_dart_image});
+    }));
+  });
+
+  auto gpu_task = fml::MakeCopyable([gpu_task_runner, picture, picture_bounds,
+                                     snapshot_delegate, ui_task]() {
+    gpu_task_runner->PostTask([snapshot_delegate, picture, picture_bounds,
+                               ui_task]() {
+      // Snapshot the picture on the GPU thread. This thread has access to the
+      // GPU contexts that may contain the sole references to a texture backed
```
Comment (jason-simmons) [inline lib/ui/compositing/scene.cc]: typo: "a texture images" should be "texture images"

**Comment by chinmaygarde** [comment]
I'll follow up with @brianosman to figure out if the texture stealing API in Skia can be made explicit so that we can transfer the texture backed image to the IO thread for device to host transfer.

**Comment by donarn** [comment]
when can we expect this fix to be available on the beta channel?

**Comment by chinmaygarde** [comment]
@cbracken Can you comment on the roll to the beta channel? This is already on ToT (was rolled in https://github.com/flutter/flutter/commit/f7f72211b68a06608eebd04be207e9ee112ac445)

**Comment by cbracken** [comment]
@donarn we don't make [forward-looking statements](https://en.wikipedia.org/wiki/Forward-looking_statement) but I've adjusted the [milestone](https://github.com/flutter/flutter/milestone/33) for flutter/flutter#17687 to accurately reflect the set of issues that fall into the same release.
